src/engine/readaheadmanager.cpp

```diff
@@ -23,88 +25,113 @@ ReadAheadManager::ReadAheadManager(CachingReader* pReader,
-            preloop_samples = samples_available;
-            samples_needed = math_clamp<SINT>(samples_needed, 0, samples_available);
+            // We can only read whole frames from the reader.
+            // Use ceil here, to be sure to reach the loop trigger.
+            preloop_samples = SampleUtil::ceilPlayPosToFrameStart(samplesToLoopTrigger,
+                    kNumChannels);
+            // clamp requested samples from the caller to the loop trigger point
+            samples_from_reader = math_clamp(requested_samples,
+                    static_cast<SINT>(0), preloop_samples);
         }
     }
 
-    if (in_reverse) {
-        start_sample = m_iCurrentPosition - samples_needed;
-    }
-
     // Sanity checks.
-    if (samples_needed < 0) {
+    if (samples_from_reader < 0) {
         qDebug() << "Need negative samples in ReadAheadManager::getNextSamples. Ignoring read";
         return 0;
     }
 
-    SINT samples_read = m_pReader->read(start_sample, in_reverse, samples_needed,
-                                       base_buffer);
+    SINT start_sample = SampleUtil::roundPlayPosToFrameStart(
+            m_currentPosition, kNumChannels);
+
+    SINT samples_read = m_pReader->read(
+            start_sample, in_reverse, samples_from_reader, pOutput);
 
-    if (samples_read != samples_needed) {
-        qDebug() << "didn't get what we wanted" << samples_read << samples_needed;
+    if (samples_read != samples_from_reader) {
+        qDebug() << "didn't get what we wanted" << samples_read << samples_from_reader;
     }
 
     // Increment or decrement current read-ahead position
+    // Mixing int and double here is desired, because the fractional frame should
+    // be resist
     if (in_reverse) {
-        addReadLogEntry(m_iCurrentPosition, m_iCurrentPosition - samples_read);
-        m_iCurrentPosition -= samples_read;
+        addReadLogEntry(m_currentPosition, m_currentPosition - samples_read);
+        m_currentPosition -= samples_read;
     } else {
-        addReadLogEntry(m_iCurrentPosition, m_iCurrentPosition + samples_read);
-        m_iCurrentPosition += samples_read;
+        addReadLogEntry(m_currentPosition, m_currentPosition + samples_read);
+        m_currentPosition += samples_read;
     }
 
     // Activate on this trigger if necessary
     if (loop_active) {
         // LoopingControl makes the decision about whether we should loop or
         // not.
-        const double loop_target = m_pLoopingControl->
-                process(dRate, m_iCurrentPosition, 0, 0);
+        const double loop_target = m_pLoopingControl->process(
+                dRate, m_currentPosition, 0, 0);
 
         if (loop_target != kNoTrigger) {
-            m_iCurrentPosition = loop_target;
+            m_currentPosition = loop_target;
+            if (preloop_samples > 0) {
+                // we are up to one frame ahead of the loop trigger
+                double overshoot = preloop_samples - samplesToLoopTrigger;
```

Comment:
DEBUG_ASSERT(preloop_samples >= samplesToLoopTrigger);
double overshoot = preloop_samples - samplesToLoopTrigger;
// we are up to one frame ahead of the loop trigger
DEBUG_ASSERT(overshoot < kNumChannels);